tests/phpunit/api/v4/Action/SaveTest.php

```diff
@@ -29,79 +28,93 @@
-        ['first_name' => 'Two', 'last_name' => 'Test', 'external_identifier' => 'def'],
-      ],
-      [
-        ['first_name' => 'One', 'last_name' => 'Changed', 'external_identifier' => 'ghi'],
-        ['first_name' => 'Two', 'last_name' => 'Changed', 'external_identifier' => 'def'],
-      ],
-      [
-        // Original insert
-        ['first_name' => 'One', 'last_name' => 'Test', 'external_identifier' => 'abc'],
-        // Match+update
-        ['first_name' => 'Two', 'last_name' => 'Changed', 'external_identifier' => 'def'],
-        // Subsequent insert
-        ['first_name' => 'One', 'last_name' => 'Changed', 'external_identifier' => 'ghi'],
-      ],
-    ];
-    // Test that we get a match on an empty string (eg. external_identifier => '')
-    $data[] = [
-      ['first_name', 'last_name', 'external_identifier'],
-      [
-        ['first_name' => 'One', 'last_name' => 'Test', 'external_identifier' => 'abc'],
-        ['first_name' => 'Two', 'last_name' => 'Test', 'external_identifier' => ''],
-      ],
-      [
-        ['first_name' => 'One', 'last_name' => 'Test', 'external_identifier' => 'ghi'],
-        ['first_name' => 'Two', 'last_name' => 'Test', 'external_identifier' => ''],
-      ],
-      [
-        // Original insert
-        ['first_name' => 'One', 'last_name' => 'Test', 'external_identifier' => 'abc'],
-        // Match+update
-        ['first_name' => 'Two', 'last_name' => 'Test', 'external_identifier' => ''],
-        // Subsequent insert
-        ['first_name' => 'One', 'last_name' => 'Test', 'external_identifier' => 'ghi'],
-      ],
-    ];
-    return $data;
+      self::assertNotEquals($bobOriginal['nick_name'], $testValue);
+
+      $anotherBob = [
+        'nick_name' => $testValue,
+        'first_name' => $bobOriginal['first_name'],
+        'last_name' => 'Changed',
+      ];
+
+      $anotherBob['id'] = Contact::save(FALSE)
+        ->setMatch(['first_name', 'nick_name'])
+        ->setRecords([$anotherBob])
+        ->execute()->single()['id'];
+
+      self::assertGreaterThan($bobOriginal['id'], $anotherBob['id']);
+
+      $allContactIds = [
+        $zakOriginal['id'],
+        $sameZakWithChangedLastName['id'],
+        $bobOriginal['id'],
+        $anotherBob['id'],
+      ];
+
+      $allCreatedAndSavedContacts = Contact::get(FALSE)
+        ->setSelect(['id', 'first_name', 'last_name', 'nick_name'])
+        ->addWhere('id', 'IN', $allContactIds)
+        ->execute()->indexBy('id');
+
+      self::assertCount(3, $allCreatedAndSavedContacts);
+
+      $expected = [
+        $zakOriginal['id'] => $sameZakWithChangedLastName,
+        $bobOriginal['id'] => $bobOriginal,
+        $anotherBob['id'] => $anotherBob,
+      ];
+
+      self::assertEquals($expected, $allCreatedAndSavedContacts->getArrayCopy());
+
+      Contact::delete(FALSE)
+        ->setUseTrash(FALSE)
+        ->addWhere('id', 'IN', $allContactIds)
+        ->execute();
```

Comment:
It was thoughtful of you to add this cleanup at the end, but that's not how we do test cleanup :)
For one thing, it's a pain for test authors to remember to do this every time, and, more importantly, if the test fails then it will never get to this step!

Best practice these days is to use `$this->createTestRecord()` instead of the api. As the name implies, the function creates test records... but it also automatically cleans them up afterwards, even if the test crashes halfway through!